### PR TITLE
Heat equation tutorial with pyamrex

### PR DESCRIPTION
**Pull request by EZoni**
Thank you to @RevathiJambunathan, @oshapoval, @dpgrote, @kngott and @ax3l for the teamwork.

Rewrite the "HeatEquation" guided tutorial using [pyamrex](https://github.com/AMReX-Codes/pyamrex).

This depends on:
- https://github.com/AMReX-Codes/pyamrex/pull/150
- https://github.com/AMReX-Codes/pyamrex/pull/153
- https://github.com/AMReX-Codes/pyamrex/pull/154
- https://github.com/AMReX-Codes/pyamrex/pull/155

To-do:
- [x] Validate results (pyamrex vs C++)
- [x] Generalize `numpy` with `cupy` (CPU -> CPU, CUDA, HIP)

Follow-up ideas:
- [ ] Use [ParmParse](https://github.com/AMReX-Codes/pyamrex/blob/development/tests/test_parmparse.py) to read input parameters
- [ ] Make code dimension-agnostic as original C++ tutorial

Visualization of plotfile data produced with pyamrex, using yt:

heat-equation-phi

Comparison of plotfile data produced with pyamrex and C++, using ParaView, at step=100:

<p align="center">
    
    
</p>

### CODE REVIEW

File: GuidedTutorials/HeatEquation/Source/main.py [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+                       -2*phiOld[:, ngz  :-ngz     , ngy  :-ngy     , ngx  :-ngx     ]
+                         +phiOld[:, ngz-1:hiz-ngz-1, ngy  :-ngy     , ngx  :-ngx     ]) / dx[2]**2))
+
+        # Update time
+        time = time + dt
+
+        # Copy new solution into old solution
+        phi_old.copy(phi_old, phi_new, 0, 0, 1, 0)
+        # TODO Use keyword arguments
+        # phi_old.copy(dst=phi_old, src=phi_new, srccomp=0, dstcomp=0, numcomp=1, nghost=0)
```
Comment (EZoni) [inline GuidedTutorials/HeatEquation/Source/main.py:113]:
```suggestion
        phi_old.copy(dst=phi_old, src=phi_new, srccomp=0, dstcomp=0, numcomp=1, nghost=0)
```

File: GuidedTutorials/HeatEquation/Source/main.py [hunk 2]
```diff
@@ -0,0 +1,141 @@
+import numpy as np
+import amrex.space3d as amr
+
+def main_main ():
+    """
+    TODO Add documentation
+    """
```
Comment (ax3l) [inline GuidedTutorials/HeatEquation/Source/main.py:7]:
```suggestion
def main ():
    """
    The main function, automatically called below if called as a script.
    """
```

File: GuidedTutorials/HeatEquation/Source/main.py [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+# size of each box (or grid)
+max_grid_size = 16
+# total steps in simulation
+nsteps = 1000
+# how often to write a plotfile
+plot_int = 100
+# time step
+dt = 1e-5
+
+main_main()
```
Comment (ax3l) [inline GuidedTutorials/HeatEquation/Source/main.py:138]: For now, do you want to make those arguments (with default values) to `main()`?

File: GuidedTutorials/HeatEquation/Source/main.py [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+        # Tell the I/O Processor to write out which step we're doing
+        print(f'Advanced step {step}\n')
+
+        # Write a plotfile of the current data (plot_int was defined in the inputs file)
+        if plot_int > 0 and step%plot_int == 0:
+            pltfile = amr.concatenate("plt", step, 5)
+            varnames = amr.Vector_string(['phi'])
+            amr.write_single_level_plotfile(pltfile, phi_new, varnames, geom, time, step)
+
+# Initialize AMReX
```
Comment (ax3l) [inline GuidedTutorials/HeatEquation/Source/main.py:122]: Wrap this all in a
```py
if __name__ == "__main__":
    main()
```
?

File: GuidedTutorials/HeatEquation/Source/main.py [hunk 5]
```diff
@@ -0,0 +1,141 @@
+import numpy as np
```
Comment (ax3l) [inline GuidedTutorials/HeatEquation/Source/main.py:1]: Sincethis can be called as a script, let's add
```suggestion
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

# Copyright 2023 The AMReX Community
#
# This file is part of AMReX.
#
# License: BSD-3-Clause-LBNL
# Authors: Revathi Jambunathan, Edoardo Zoni, Olga Shapoval, Axel Huebl

import numpy as np
```

and also change file permissions via `chmod a+rx main.py` to be executable?

File: GuidedTutorials/HeatEquation/Source/main.py [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+    ng = phi_old.nGrowVect
+    ngx = ng[0]
+    ngy = ng[1]
+    ngz = ng[2]
+
+    # Loop over boxes
+    for mfi in phi_old:
+        bx = mfi.validbox()
+        # phiOld is indexed in reversed order (z,y,x) and indices are local
+        phiOld = np.array(phi_old.array(mfi), copy=False)
```
Comment (ax3l) [inline GuidedTutorials/HeatEquation/Source/main.py:65]: @RevathiJambunathan are you still pushing the `cupy`/`numpy` generalization we worked on last Friday? :)

**Comment by ax3l** [comment]
reopening for CI update from #97

File: GuidedTutorials/HeatEquation/Source/main.py [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+                         +phiOld[:, ngz-1:hiz-ngz-1, ngy  :-ngy     , ngx  :-ngx     ]) / dx[2]**2))
+
+        # Update time
+        time = time + dt
+
+        # Copy new solution into old solution
+        phi_old.copy(dst=phi_old, src=phi_new, srccomp=0, dstcomp=0, numcomp=1, nghost=0)
+
+        # Tell the I/O Processor to write out which step we're doing
+        print(f'Advanced step {step}\n')
```
Comment (atmyers) [inline GuidedTutorials/HeatEquation/Source/main.py:114]: I think this will print on all procs, no? Is `amrex::Print()` or something like it exposed yet?

Comment (ax3l) [reply GuidedTutorials/HeatEquation/Source/main.py:114]: I think we have not exposed an `amrex::Print()` equivalent, but the IOProcessor is exposed for an `if` to be used in here:
https://github.com/AMReX-Codes/pyamrex/blob/development/src/Base/ParallelDescriptor.cpp

Comment (ax3l) [reply GuidedTutorials/HeatEquation/Source/main.py:114]: PR for `amr.Print(...)` in https://github.com/AMReX-Codes/pyamrex/pull/174

File: GuidedTutorials/HeatEquation/Source/main.py (hunk 6 above)
Comment (ax3l) [reply GuidedTutorials/HeatEquation/Source/main.py]: I pushed this now :)

**Comment by ax3l** [comment]
@EZoni careful, I rebased this on `main` to include updates from #95

**Comment by ax3l** [comment]
@EZoni @atmyers @RevathiJambunathan this is also ready for review :)

File: GuidedTutorials/HeatEquation/Source/main.py (hunk 7 above)
Comment (ax3l) [reply GuidedTutorials/HeatEquation/Source/main.py]: added

**Comment by ax3l** [comment]
@EZoni @atmyers @RevathiJambunathan ping :)

**Pull request merged**